standard/expressions.md

```diff
@@ -2370,6 +2371,77 @@ A *default_value_expression* is a constant expression ([§12.20](expressions.md#
 ### 12.7.16 Anonymous method expressions
 
 An *anonymous_method_expression* is one of two ways of defining an anonymous function. These are further described in [§12.16](expressions.md#1216-anonymous-function-expressions).
+	
+### 12.7(.x) §stack-allocation Stack allocation
+
+A stack allocation initializer allocates a block of memory from the call stack.
+
+```ANTLR
+stackalloc_initializer
+    : 'stackalloc' unmanaged_type '[' expression ']'
+    | 'stackalloc' unmanaged_type? '[' expression? ']' stackalloc_initializer_elements
+    ;
+
+stackalloc_initializer_elements
+    : '{' stackalloc_initializer_element_list? '}'
+    | '{' stackalloc_initializer_element_list ',' '}'
+    ;
+
+stackalloc_initializer_element_list
+    : stackalloc_element_initializer (',' stackalloc_element_initializer)*
+    ;
+    
+stackalloc_element_initializer
+    : expression
+    ;
+```
+
+The *unmanaged_type* ([§9.8](types.md#98-unmanaged-types)) indicates the type of the items that will be stored in the newly allocated location, and the *expression* indicates the number of these items. Taken together, these specify the required allocation size. As the size of a stack allocation cannot be negative, it is a compile-time error to specify the number of items as a *constant_expression* that evaluates to a negative value.
+
+If *unmanaged_type* is omitted, it is inferred from the corresponding *stackalloc_initializer_elements*. If *expression* is omitted from *stackalloc_initializer*, it is inferred to be the number of *stackalloc_element_initializer*s in the corresponding *stackalloc_initializer_elements*.
+
+When a *stackalloc_initializer* includes both *expression* and *stackalloc_initializer_elements* the number of elements in that *stackalloc_initializer_elements* shall match the value of *expression*.
+
+A stack allocation initializer of the form `stackalloc T[E]` requires `T` to be an *unmanaged_type* and `E` to be an expression implicitly convertible to type `int`. The operator allocates `E * sizeof(T)` bytes from the call stack, and the resulting type and value are determined, as follows:
+
+- If *unmanaged_type* is a *pointer_type* ([§23.3](unsafe-code.md#233-pointer-types)) or if *stackalloc_initializer* is used in a context that requires a *pointer_type*, the result is a pointer, of type `T*`, to the newly allocated block. If the context cannot be inferred, a pointer context is assumed. As pointer contexts require unsafe code, see [§23.3](unsafe-code.md#239-stack-allocation) for more information.
+- Otherwise, the result has type `System.Span<T>` and maps to the newly allocated block. Apart from being used as the initializer of a local variable, this result shall be permitted in the following contexts only:
+  - The right operand of an *assignment_operator* that is not embedded in some larger expression
+  - The second and third operands in a *conditional_expression*
```

Comment:
Seems like this would be invalid then:

```C#
using System;
public class C {
    public void M() {
        _ = ((Span<byte>)stackalloc byte[1])[0];
    }
}
```

Reply:
Oh right, it is invalid in C# 7, but not in C# 8.

Reply:
The proposed wording would allow the following *stackalloc_initializer* because it is the second operand in a *conditional_expression*, yet the compiler rejects it in C# 7.

```C#
using System;
public class C {
    public void M() {
        _ = ((Span<byte>)(true ? stackalloc byte[1] : default(Span<byte>)))[0];
    }
}
```

Reply:
*assignment_operator* includes compound assignment operators, so the wording would allow this as well:

```C#
using System;
public class C {
    public static C operator +(C c, Span<byte> s) { return c; }
    public void M(C c) {
        c += stackalloc byte[1];
    }
}
```

But the compiler rejects it in C# 7.

Reply:
If *stackalloc_initializer* can be used in *assignment_operator* or *conditional_expression*, then that needs some changes in the grammar too. So far, *stackalloc_initializer* is referenced only from *local_variable_initializer*.

Reply:
> Oh right, it is invalid in C# 7, but valid in C# 8.

C# 7 seems the wiser choice here, why is it valid in C# 8?

Reply:
> The proposed wording would allow the following _stackalloc_initializer_ because it is the second operand in a _conditional_expression_, yet the compiler rejects it in C# 7.
> 
> ```cs
> using System;
> public class C {
>     public void M() {
>         _ = ((Span<byte>)(true ? stackalloc byte[1] : default(Span<byte>)))[0];
>     }
> }
> ```

Was the text meant to imply “The second __*and*__ third operands in a”, i.e. neither or both branches? One could imagine a reason for that restriction.

Reply:
Admittedly, my head hurts now. 

The previous example fails because the `stackalloc` is part of a "nested expression". The following variation compiles in C# 7.3:

```csharp
_ = true ? stackalloc byte[1] : default(Span<byte>);
```

I'm asking for a definition of "nested expression". It doesn't occur in the standard, and I can't find the term used in LDM notes.

Reply:
This answer turns out to simplify the PR somewhat. In C# 7.3, a *stack_initializer* can be used only as a *local_variable_initializer*. In that context, the result is `T*`, with an implicit conversation to `Span<T>` for use in safe contexts. 

That means the `stackalloc` initializer can be used as *either or both* the 2nd and 3rd operands of a *conditional_expression*.